mods/tnt/init.lua

```diff
@@ -160,22 +155,54 @@ local function add_effects(pos, radius)
 		maxsize = 16,
 		texture = "tnt_smoke.png",
 	})
+
+	-- we just dropped some items. Look at the items entities and pick
+	-- one of them to use as texture
+	local texture = "tnt_blast.png" --fallback texture
+	local objs = minetest.get_objects_inside_radius(pos, 2)
+	for _, obj in pairs(objs) do
+		if obj and obj:get_luaentity() then
+			local def = ItemStack(obj:get_luaentity().itemstring):get_definition()
+			if def.tiles then
+				texture = def.tiles[1]
+				break
+			end
+		end
+	end
```

Comment:
Should do something like this for performance reasons and so you don't use a texture from something you didn't destroy:
```Lua
for _, item in pairs(drops) do
	local def = item.get_definition()
	...
end
```

You might also consider doing something like trying the most destroyed nodes first, so you don't get showered with dirt instead of stone if your cave happens t have one or two nodes of dirt.